plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,221 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
 	return sortBinlogFiles(binlogFiles), nil
 }
 
+// ConvertTsToLocalBinlogCoordinate converts a timestamp to binlog coordinate using local binlog files.
+// We do all kinds of comparison like finding proper backup using the converted BinlogInfo, but not the original timestamp.
+func (r *Restore) ConvertTsToLocalBinlogCoordinate(ctx context.Context, targetTs int64) (*api.BinlogInfo, error) {
+	binlogFilesLocalSorted, err := GetSortedLocalBinlogFiles(r.binlogDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read sorted local binlog files, error: %w", err)
+	}
+	if !binlogFilesAreContinuous(binlogFilesLocalSorted) {
+		return nil, fmt.Errorf("local binlog files are not continuous")
+	}
+
+	var firstEventTsListSorted []int64
+	for _, file := range binlogFilesLocalSorted {
+		eventTs, err := r.parseFirstLocalBinlogEventTs(ctx, file.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse the local binlog file %q's first binlog event ts, error: %w", file.Name, err)
+		}
+		firstEventTsListSorted = append(firstEventTsListSorted, eventTs)
+	}
+	binlogFileTargetIndex, err := getLastTsIndexBeforeTargetTs(firstEventTsListSorted, targetTs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find the local binlog file containing targetTs %d, error: %w", targetTs, err)
+	}
+	binlogFileTarget := binlogFilesLocalSorted[binlogFileTargetIndex]
+	eventStartPos, err := r.getFirstBinlogEventStartPosAfterTs(ctx, binlogFileTarget, targetTs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find the binlog event after targetTs %d, error: %w", targetTs, err)
+	}
+	return &api.BinlogInfo{FileName: binlogFileTarget.Name, Position: eventStartPos}, nil
+}
+
+func parseBinlogEventTsInLine(line string) (eventTs int64, found bool, err error) {
+	// The target line starts with string like "#220421 14:49:26 server id 1"
+	if !strings.Contains(line, "server id") {
+		return 0, false, nil
+	}
+	if strings.Contains(line, "end_log_pos 0") {
+		// https://github.com/mysql/mysql-server/blob/8.0/client/mysqlbinlog.cc#L1209-L1212
+		// Fake events with end_log_pos=0 could be generated and we need to ignore them.
+		return 0, false, nil
+	}
+	fields := strings.Fields(line)
+	// fields should starts with ["#220421", "14:49:26", "server", "id", "1", "end_log_pos", "34794"]
+	if len(fields) < 7 ||
+		(len(fields[0]) != 7 || len(fields[1]) != 8 || fields[2] != "server" || fields[3] != "id" || fields[5] != "end_log_pos") {
+		return 0, false, fmt.Errorf("found unexpected mysqlbinlog output line %q when parsing binlog event timestamp and end position", line)
+	}
+	datetime, err := time.ParseInLocation("060102 15:04:05", fmt.Sprintf("%s %s", fields[0][1:], fields[1]), time.Local)
+	if err != nil {
+		return 0, false, err
+	}
+	return datetime.Unix(), true, nil
+}
+
+func parseBinlogEventStartPosInLine(line string) (startPos int64, found bool, err error) {
```

Comment:
Don't mention StartPosition because there's only Position throughout the context. There's no end position.

Reply:
SG.